kafka/src/main/java/org/apache/apex/malhar/kafka/KafkaSinglePortExactlyOnceOutputOperator.java

```diff
@@ -0,0 +1,369 @@
+  {
+    String[] split = key.split(KEY_SEPARATOR);
+
+    if (split.length != 2) {
+      return false;
+    }
+
+    if ((Integer.parseInt(split[1]) == operatorId) && (split[0].equals(appName))) {
+      return true;
+    }
+
+    return false;
+  }
+
+  private boolean alreadyInKafka(T message)
+  {
+    if ( windowId <= windowDataManager.getLargestRecoveryWindow() ) {
+      return true;
+    }
+
+    if (partialWindowTuples.containsKey(message)) {
+
+      Integer val = partialWindowTuples.get(message);
+
+      if ( val == 0 ) {
+        return false;
+      } else if ( val == 1 ) {
+        partialWindowTuples.remove(message);
+      } else {
+        partialWindowTuples.put(message, val - 1);
+      }
+
+      return true;
+    }
+
+    return false;
+  }
+
+  private Map<Integer,Long> getPartitionsAndOffsets() throws ExecutionException, InterruptedException
+  {
+    List<PartitionInfo> partitionInfoList = consumer.partitionsFor(getTopic());
+    List<TopicPartition> topicPartitionList = new java.util.ArrayList<>();
+
+    for ( PartitionInfo partitionInfo: partitionInfoList) {
+      topicPartitionList.add(new TopicPartition(getTopic(), partitionInfo.partition()) );
+    }
+
+    Map<Integer,Long> parttionsAndOffset = new HashMap<>();
+    consumer.assign(topicPartitionList);
+
+    for ( PartitionInfo partitionInfo: partitionInfoList) {
+
+      try {
+
+        TopicPartition topicPartition = new TopicPartition(getTopic(), partitionInfo.partition());
+        consumer.seekToEnd(topicPartition);
+        parttionsAndOffset.put(partitionInfo.partition(), consumer.position(topicPartition));
+
+      } catch (Exception ex) {
+        throw new RuntimeException(ex);
+      }
+    }
+
+    return parttionsAndOffset;
+  }
+
+  private void rebuildPartialWindow()
+  {
+    logger.info("Rebuild the partial window after " + windowDataManager.getLargestRecoveryWindow());
+
+    Map<Integer,Long> storedOffsets = getStoredOffsets();
+    Map<Integer,Long> currentOffsets = getCurrentOffsets();
+
+    if (storedOffsets == null || currentOffsets == null) {
+      logger.debug("No tuples found while building partial window " + windowDataManager.getLargestRecoveryWindow());
+      return;
+    }
+
+    List<TopicPartition> topicPartitions = new ArrayList<>();
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      topicPartitions.add(new TopicPartition(getTopic(), entry.getKey()));
+    }
+
+    consumer.assign(topicPartitions);
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      Long storedOffset = 0L;
+      Integer currentPartition = entry.getKey();
+      Long currentOffset = entry.getValue();
+
+      if (storedOffsets.containsKey(currentPartition)) {
+        storedOffset = storedOffsets.get(currentPartition);
+      }
+
+      if (storedOffset >= currentOffset) {
+        continue;
+      }
+
+      consumer.seek(new TopicPartition(getTopic(), currentPartition), storedOffset);
```

Comment:
This line could possibly throw an exception if the storedOffset is invalid because of corrupted windowed data or maybe the messages are purged from kafka. We can log an warn line saying "exactly once" can't be guaranteed and seek to the end of the queue